app/src/main/java/com/amaze/filemanager/ui/dialogs/OpenFileDialogFragment.kt

```diff
@@ -0,0 +1,379 @@
+import com.amaze.filemanager.ui.activities.superclasses.BasicActivity
+import com.amaze.filemanager.ui.activities.superclasses.PreferenceActivity
+import com.amaze.filemanager.ui.activities.superclasses.ThemedActivity
+import com.amaze.filemanager.ui.base.BaseBottomSheetFragment
+import com.amaze.filemanager.ui.icons.MimeTypes
+import com.amaze.filemanager.ui.provider.UtilitiesProvider
+import com.amaze.filemanager.ui.views.ThemedTextView
+
+class OpenFileDialogFragment : BaseBottomSheetFragment() {
+
+    private var uri: Uri? = null
+    private var mimeType: String? = null
+    private var useNewStack: Boolean? = null
+
+    private lateinit var viewBinding: FragmentOpenFileDialogBinding
+    private lateinit var adapter: AppsAdapter
+    private lateinit var utilsProvider: UtilitiesProvider
+    private lateinit var sharedPreferences: SharedPreferences
+
+    companion object {
+
+        private const val KEY_URI = "uri"
+        private const val KEY_MIME_TYPE = "mime_type"
+        private const val KEY_USE_NEW_STACK = "use_new_stack"
+        private const val KEY_PREFERENCES_DEFAULT = "_DEFAULT"
+        const val KEY_PREFERENCES_LAST = "_LAST"
+
+        /**
+         * Opens the file using previously set default app or shows a bottom sheet dialog
+         */
+        fun openFileOrShow(
+            uri: Uri,
+            mimeType: String,
+            useNewStack: Boolean,
+            activity: PreferenceActivity,
+            forceChooser: Boolean
+        ) {
+            if (mimeType == "*/*" || forceChooser || !getPreferenceAndStartActivity(
+                    uri, mimeType, useNewStack, activity
+                )
+            ) {
+                if (forceChooser) {
+                    clearMimeTypePreference(
+                        MimeTypes.getMimeType(uri.toString(), false), activity.prefs
+                    )
+                }
+                val openFileDialogFragment = newInstance(uri, mimeType, useNewStack)
+                openFileDialogFragment.show(
+                    activity.supportFragmentManager, javaClass.simpleName
+                )
+            }
+        }
+
+        private fun newInstance(uri: Uri, mimeType: String, useNewStack: Boolean):
+            OpenFileDialogFragment {
+                val args = Bundle()
+
+                val fragment = OpenFileDialogFragment()
+                args.putParcelable(KEY_URI, uri)
+                args.putString(KEY_MIME_TYPE, mimeType)
+                args.putBoolean(KEY_USE_NEW_STACK, useNewStack)
+                fragment.arguments = args
+                return fragment
+            }
+
+        private fun startActivity(context: Context, intent: Intent) {
+            try {
+                context.startActivity(intent)
+            } catch (e: ActivityNotFoundException) {
+                Log.e(javaClass.simpleName, e.message, e)
+                Toast.makeText(context, R.string.no_app_found, Toast.LENGTH_SHORT).show()
+                throw e
```

Comment:
Throw with no Throws annotation, my recomendation would be to make UI errors silent (or just a toast) and file errors crashy.

Reply:
This is caught later on in `getPreferenceAndStartActivity`  in case some app which was earlier set default was uninstalled. We try to run with it and if that activity is not found then we show the chooser sheet again.

Reply:
The Throws annotation is for the compiler to catch issues with non catched exceptions. To make sure each use case of this function catches this exception.

Reply:
Throws still not forcing compiler to check the exception in kotlin.

Reply:
Okay.